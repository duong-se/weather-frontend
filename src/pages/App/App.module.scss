.root {
  background-color: #ebedf0;
  min-height: 100vh;
  .information {
    margin: 0 auto;
  }
  .weatherWidget {
    min-height: 30rem;
    max-width: 55rem;
    margin: 0 auto;
  }
  .temperatureText {
    font-size: 5rem;
  }
}
