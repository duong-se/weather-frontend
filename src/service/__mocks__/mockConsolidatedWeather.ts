export const mockConsolidatedWeather = {
  "id": 6547671268458496,
  "weather_state_name": "Light Cloud",
  "weather_state_abbr": "lc",
  "wind_direction_compass": "SW",
  "created": "2020-09-22T15:20:01.480287Z",
  "applicable_date": "2020-09-22",
  "min_temp": 14.78,
  "max_temp": 24,
  "the_temp": 24.035,
  "wind_speed": 5.706272015681752,
  "wind_direction": 217.66164574425014,
  "air_pressure": 1008.5,
  "humidity": 51,
  "visibility": 9.400414081762506,
  "predictability": 70
}
