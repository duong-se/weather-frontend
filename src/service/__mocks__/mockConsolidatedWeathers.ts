export const mockConsolidatedWeathers = [
  {
    "id": 6547671268458496,
    "weather_state_name": "Light Cloud",
    "weather_state_abbr": "lc",
    "wind_direction_compass": "SW",
    "created": "2020-09-22T15:20:01.480287Z",
    "applicable_date": "2020-09-22",
    "min_temp": 14.78,
    "max_temp": 24,
    "the_temp": 24.035,
    "wind_speed": 5.706272015681752,
    "wind_direction": 217.66164574425014,
    "air_pressure": 1008.5,
    "humidity": 51,
    "visibility": 9.400414081762506,
    "predictability": 70
  },
  {
    "id": 4713030018400256,
    "weather_state_name": "Heavy Rain",
    "weather_state_abbr": "hr",
    "wind_direction_compass": "SW",
    "created": "2020-09-22T15:20:02.471805Z",
    "applicable_date": "2020-09-23",
    "min_temp": 11.469999999999999,
    "max_temp": 19.13,
    "the_temp": 18.705,
    "wind_speed": 7.6319694457806415,
    "wind_direction": 229.33431218602843,
    "air_pressure": 1001.5,
    "humidity": 70,
    "visibility": 8.366141732283463,
    "predictability": 77
  },
  {
    "id": 6312566033743872,
    "weather_state_name": "Heavy Rain",
    "weather_state_abbr": "hr",
    "wind_direction_compass": "SW",
    "created": "2020-09-22T15:20:02.052900Z",
    "applicable_date": "2020-09-24",
    "min_temp": 8.844999999999999,
    "max_temp": 15.34,
    "the_temp": 15.274999999999999,
    "wind_speed": 9.736295492422917,
    "wind_direction": 225.3331262999033,
    "air_pressure": 996.5,
    "humidity": 59,
    "visibility": 14.177205122087011,
    "predictability": 77
  },
  {
    "id": 4546612517928960,
    "weather_state_name": "Light Rain",
    "weather_state_abbr": "lr",
    "wind_direction_compass": "NNW",
    "created": "2020-09-22T15:20:01.658749Z",
    "applicable_date": "2020-09-25",
    "min_temp": 7.495,
    "max_temp": 13.82,
    "the_temp": 13.04,
    "wind_speed": 10.535621595103263,
    "wind_direction": 334.5010293182893,
    "air_pressure": 1004.5,
    "humidity": 61,
    "visibility": 14.033979062276305,
    "predictability": 75
  },
  {
    "id": 4604599405117440,
    "weather_state_name": "Clear",
    "weather_state_abbr": "c",
    "wind_direction_compass": "NNW",
    "created": "2020-09-22T15:20:02.569048Z",
    "applicable_date": "2020-09-26",
    "min_temp": 7.385,
    "max_temp": 13.649999999999999,
    "the_temp": 13.485,
    "wind_speed": 8.837651108873512,
    "wind_direction": 330.16248530421626,
    "air_pressure": 1015,
    "humidity": 48,
    "visibility": 17.28157559850473,
    "predictability": 68
  },
  {
    "id": 6338147862446080,
    "weather_state_name": "Clear",
    "weather_state_abbr": "c",
    "wind_direction_compass": "NW",
    "created": "2020-09-22T15:20:04.446373Z",
    "applicable_date": "2020-09-27",
    "min_temp": 5.655,
    "max_temp": 14.48,
    "the_temp": 12.6,
    "wind_speed": 4.190453948938201,
    "wind_direction": 314.5,
    "air_pressure": 1021,
    "humidity": 47,
    "visibility": 9.999726596675416,
    "predictability": 68
  }
]