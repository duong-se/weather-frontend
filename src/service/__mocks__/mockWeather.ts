export const mockWeather = {
  "consolidated_weather": [
    {
      "air_pressure": 1009.5,
      "applicable_date": "2020-09-23",
      "created": "2020-09-23T16:00:46.374603Z",
      "humidity": 71,
      "id": 5728148488454144,
      "max_temp": 33.65,
      "min_temp": 25.015,
      "predictability": 77,
      "the_temp": 31.549999999999997,
      "visibility": 8.473638948540524,
      "weather_state_abbr": "hr",
      "weather_state_name": "Heavy Rain",
      "wind_direction": 157.83272289145478,
      "wind_direction_compass": "SSE",
      "wind_speed": 4.518443377942151
    },
    {
      "air_pressure": 1011,
      "applicable_date": "2020-09-24",
      "created": "2020-09-23T16:00:49.298547Z",
      "humidity": 72,
      "id": 6062055889043456,
      "max_temp": 33.5,
      "min_temp": 24.759999999999998,
      "predictability": 77,
      "the_temp": 31.44,
      "visibility": 9.714517219438479,
      "weather_state_abbr": "hr",
      "weather_state_name": "Heavy Rain",
      "wind_direction": 231.9007449072647,
      "wind_direction_compass": "SW",
      "wind_speed": 4.740631533272356
    },
    {
      "air_pressure": 1009.5,
      "applicable_date": "2020-09-25",
      "created": "2020-09-23T16:00:52.504367Z",
      "humidity": 69,
      "id": 6203842426306560,
      "max_temp": 33.3,
      "min_temp": 25.36,
      "predictability": 75,
      "the_temp": 31.92,
      "visibility": 13.932074186749384,
      "weather_state_abbr": "lr",
      "weather_state_name": "Light Rain",
      "wind_direction": 281.48523379083207,
      "wind_direction_compass": "WNW",
      "wind_speed": 6.146027525416899
    },
    {
      "air_pressure": 1009,
      "applicable_date": "2020-09-26",
      "created": "2020-09-23T16:00:55.494767Z",
      "humidity": 72,
      "id": 6404202113794048,
      "max_temp": 33.125,
      "min_temp": 25.115,
      "predictability": 77,
      "the_temp": 30.545,
      "visibility": 11.966366420106578,
      "weather_state_abbr": "hr",
      "weather_state_name": "Heavy Rain",
      "wind_direction": 286.4991601759091,
      "wind_direction_compass": "WNW",
      "wind_speed": 5.094320465065352
    },
    {
      "air_pressure": 1009,
      "applicable_date": "2020-09-27",
      "created": "2020-09-23T16:00:58.203379Z",
      "humidity": 80,
      "id": 5265468911779840,
      "max_temp": 29.575000000000003,
      "min_temp": 24.18,
      "predictability": 77,
      "the_temp": 28.6,
      "visibility": 9.820461007714945,
      "weather_state_abbr": "hr",
      "weather_state_name": "Heavy Rain",
      "wind_direction": 259.5963365607041,
      "wind_direction_compass": "W",
      "wind_speed": 3.8885642181765157
    },
    {
      "air_pressure": 1009,
      "applicable_date": "2020-09-28",
      "created": "2020-09-23T16:01:02.367098Z",
      "humidity": 77,
      "id": 5425837786529792,
      "max_temp": 30.035,
      "min_temp": 24.165,
      "predictability": 77,
      "the_temp": 29.56,
      "visibility": 9.536183687266364,
      "weather_state_abbr": "hr",
      "weather_state_name": "Heavy Rain",
      "wind_direction": 182.5,
      "wind_direction_compass": "S",
      "wind_speed": 2.9403026326254675
    }
  ],
  "latt_long": "10.759180,106.662498",
  "location_type": "City",
  "parent": {
    "latt_long": "15.974210,107.868042",
    "location_type": "Country",
    "title": "Vietnam",
    "woeid": 23424984
  },
  "sources": [
    {
      "crawl_rate": 360,
      "slug": "bbc",
      "title": "BBC",
      "url": "http://www.bbc.co.uk/weather/"
    },
    {
      "crawl_rate": 480,
      "slug": "forecast-io",
      "title": "Forecast.io",
      "url": "http://forecast.io/"
    },
    {
      "crawl_rate": 180,
      "slug": "met-office",
      "title": "Met Office",
      "url": "http://www.metoffice.gov.uk/"
    },
    {
      "crawl_rate": 360,
      "slug": "openweathermap",
      "title": "OpenWeatherMap",
      "url": "http://openweathermap.org/"
    },
    {
      "crawl_rate": 360,
      "slug": "world-weather-online",
      "title": "World Weather Online",
      "url": "http://www.worldweatheronline.com/"
    }
  ],
  "sun_rise": "2020-09-23T05:42:24.394809+07:00",
  "sun_set": "2020-09-23T17:48:41.332796+07:00",
  "time": "2020-09-23T23:25:11.462420+07:00",
  "timezone": "Asia/Ho_Chi_Minh",
  "timezone_name": "LMT",
  "title": "Ho Chi Minh City",
  "woeid": 1252431
}
