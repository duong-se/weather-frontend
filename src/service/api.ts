import fetch from 'isomorphic-fetch'
import { Weather, WeatherLocation } from '../typings'

export type FetchLocationParams = {
  locationName: string
}

export const ERROR_MESSAGE = {
  LOCATION_NOT_FOUND: 'Location Not Found',
  FALLBACK_ERROR: 'Error when get the weather',
}

export const HTTP_STATUS = {
  OK: 200,
  NOT_FOUND: 404,
}

export const fetchLocation = ({ locationName }: FetchLocationParams): Promise<WeatherLocation[]> => {
  return new Promise((resolve, reject) => {
    fetch(`${process.env.REACT_APP_API_URL}/api/location/search/?query=${locationName}`)
      .then((response) => {
        if (response.status !== HTTP_STATUS.OK) {
          reject({ message: ERROR_MESSAGE.FALLBACK_ERROR })
          return
        }
        if (response.status === HTTP_STATUS.OK) {
          return response.json()
        }
      })
      .then(resolve)
      .catch(reject)
  })
}

export type FetchWeatherParams = {
  woeid: number
}

export const fetchWeather = ({ woeid }: FetchWeatherParams): Promise<Weather> => {
  return new Promise((resolve, reject) => {
    if (!woeid) {
      reject({ message: ERROR_MESSAGE.LOCATION_NOT_FOUND })
      return
    }
    fetch(`${process.env.REACT_APP_API_URL}/api/location/${woeid}/`)
      .then((response) => {
        if (response.status === HTTP_STATUS.NOT_FOUND) {
          reject({ message: ERROR_MESSAGE.LOCATION_NOT_FOUND })
          return
        }
        if (response.status !== HTTP_STATUS.OK) {
          reject({ message: ERROR_MESSAGE.FALLBACK_ERROR })
          return
        }
        if (response.status === HTTP_STATUS.OK) {
          return response.json()
        }
      })
      .then(resolve)
      .catch(reject)
  })
}
