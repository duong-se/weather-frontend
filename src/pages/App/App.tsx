import React from 'react'
import Alert from '../../components/Alert/Alert'
import ForecastList from '../../components/ForecastList'
import LocationSearchForm from '../../components/LocationSearchForm'
import { fetchLocation, fetchWeather } from '../../service/api'
import { Weather } from '../../typings'
import { getDay } from '../../utils/getDay'
import { roundNumber } from '../../utils/roundNumber'
import style from './App.module.scss'

export const renderCurrentWeatherInfo = (currentWeatherIndex: number, weather?: Weather) => {
  const currentWeather = weather?.consolidated_weather?.[currentWeatherIndex]
  if (!currentWeather) {
    return null
  }
  return (
    <div className="text-center">
      <h1>{currentWeather?.weather_state_name}</h1>
      {currentWeather?.the_temp && (
        <div>
          <span className={style.temperatureText}>{roundNumber(currentWeather?.the_temp)}</span>
          <span>&#8451;</span>
        </div>
      )}
      <h4>{weather?.title ?? ''}</h4>
      <h6>{getDay(currentWeather?.applicable_date)}</h6>
    </div>
  )
}

export const renderError = (errorMessage?: string) => {
  if (!errorMessage) {
    return null
  }
  return (
    <div className="row no-gutters mb-4">
      <Alert message={errorMessage} />
    </div>
  )
}

export type OnSubmitParams = {
  setLoading: React.Dispatch<React.SetStateAction<boolean>>
  setError: React.Dispatch<React.SetStateAction<string | undefined>>
  setWeather: React.Dispatch<React.SetStateAction<Weather | undefined>>
}

export const onSubmit = ({ setLoading, setError, setWeather }: OnSubmitParams) => async (
  e: React.BaseSyntheticEvent<any>,
) => {
  e.preventDefault()
  setLoading(true)
  setError('')
  setWeather(undefined)
  const locationName = e?.target?.locationName?.value
  try {
    const locations = await fetchLocation({ locationName })
    const firstWeatherLocationWoeid = locations?.[0]?.woeid
    const weathers = await fetchWeather({ woeid: firstWeatherLocationWoeid })
    setWeather(weathers)
  } catch (error) {
    setError(error?.message)
  } finally {
    setLoading(false)
  }
}

const App: React.FC = () => {
  const [weather, setWeather] = React.useState<Weather>()
  const [isLoading, setLoading] = React.useState<boolean>(false)
  const [errorMessage, setError] = React.useState<string>()
  const [currentWeatherIndex, setCurrentWeatherIndex] = React.useState<number>(0)

  return (
    <div className={`${style.root} container-fluid`}>
      <div className={`card p-4 ${style.weatherWidget}`}>
        <div className={`row no-gutters mb-4 ${style.information}`}>
          <LocationSearchForm isLoading={isLoading} onSubmit={onSubmit({ setLoading, setError, setWeather })} />
        </div>
        {renderError(errorMessage)}
        <div className={`row no-gutters mb-4 ${style.information}`}>
          {renderCurrentWeatherInfo(currentWeatherIndex, weather)}
        </div>
        <div className={`row no-gutters ${style.information}`}>
          <ForecastList
            setCurrentWeather={setCurrentWeatherIndex}
            consolidatedWeathers={weather?.consolidated_weather ?? []}
          />
        </div>
      </div>
    </div>
  )
}

export default App
