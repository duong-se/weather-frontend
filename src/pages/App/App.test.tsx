import React from 'react'
import { shallow } from 'enzyme'
import App, { renderCurrentWeatherInfo, renderError, onSubmit } from './App'
import { Weather } from '../../typings'
import { mockWeather } from '../../service/__mocks__/mockWeather'
import { fetchLocation, fetchWeather } from '../../service/api'
import { mockLocation } from '../../service/__mocks__/mockLocation'

jest.mock('../../service/api', () => ({ fetchLocation: jest.fn(), fetchWeather: jest.fn() }))

describe('App', () => {
  describe('App', () => {
    it('should match snapshot', () => {
      const wrapper = shallow(<App />)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('renderCurrentWeatherInfo', () => {
    it('should render correctly and have data', () => {
      const wrapper = shallow(<div>{renderCurrentWeatherInfo(1, mockWeather)}</div>)
      const colWrapper = wrapper.find('[className="text-center"]')
      expect(colWrapper).toHaveLength(1)
      expect(wrapper).toMatchSnapshot()
    })

    it('should render correctly and no data', () => {
      const wrapper = shallow(<div>{renderCurrentWeatherInfo(1, {} as Weather)}</div>)
      const colWrapper = wrapper.find('[className="text-center"]')
      expect(colWrapper).toHaveLength(0)
      expect(wrapper).toMatchSnapshot()
    })

    it('should render correctly and have title', () => {
      const wrapper = shallow(<div>{renderCurrentWeatherInfo(1, mockWeather)}</div>)
      const stateText = wrapper.find('h1').text()
      expect(stateText).toEqual('Heavy Rain')
      const locationTitle = wrapper.find('h4').text()
      expect(locationTitle).toEqual('Ho Chi Minh City')
      const dayText = wrapper.find('h6').text()
      expect(dayText).toEqual('Thursday')
    })
  })

  describe('renderError', () => {
    it('should renderError', () => {
      const wrapper = shallow(<div>{renderError('MockError')}</div>)
      const alertWrapper = wrapper.find('Alert')
      expect(alertWrapper).toHaveLength(1)
      expect(wrapper).toMatchSnapshot()
    })

    it('should render null', () => {
      const wrapper = shallow(<div>{renderError('')}</div>)
      const alertWrapper = wrapper.find('Alert')
      expect(alertWrapper).toHaveLength(0)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('onSubmit', () => {
    it('should callSetLoading setError, setWeather', (done) => {
      ;(fetchLocation as jest.Mocked<any>).mockReturnValueOnce(Promise.resolve({ data: mockLocation }))
      ;(fetchWeather as jest.Mocked<any>).mockReturnValueOnce(Promise.resolve({ data: mockWeather }))
      const mockEvent = {
        preventDefault: jest.fn(),
        target: {
          locationName: {
            value: 'mockValue',
          },
        },
      } as any
      const mockSetLoading = jest.fn()
      const mockSetError = jest.fn()
      const mockSetWeather = jest.fn()
      const fn = onSubmit({ setLoading: mockSetLoading, setError: mockSetError, setWeather: mockSetWeather })
      fn(mockEvent)
      expect(mockEvent.preventDefault).toBeCalled()
      expect(mockSetLoading).toBeCalledTimes(1)
      expect(mockSetError).toBeCalledWith('')
      expect(mockSetWeather).toBeCalledTimes(1)
      setTimeout(() => {
        expect(mockSetWeather).toBeCalledTimes(2)
        expect(mockSetLoading).toBeCalledTimes(2)
        done()
      }, 1000)
    })

    it('should call only setError', (done) => {
      ;(fetchLocation as jest.Mocked<any>).mockReturnValueOnce(Promise.resolve({ message: 'mockError' }))
      ;(fetchWeather as jest.Mocked<any>).mockReturnValueOnce(Promise.reject({ message: 'mockError' }))
      const mockEvent = {
        preventDefault: jest.fn(),
        target: {
          locationName: {
            value: 'mockValue',
          },
        },
      } as any
      const mockSetLoading = jest.fn()
      const mockSetError = jest.fn()
      const mockSetWeather = jest.fn()
      const fn = onSubmit({ setLoading: mockSetLoading, setError: mockSetError, setWeather: mockSetWeather })
      fn(mockEvent)
      expect(mockEvent.preventDefault).toBeCalled()
      expect(mockSetLoading).toBeCalledTimes(1)
      expect(mockSetError).toBeCalledWith('')
      expect(mockSetWeather).toBeCalledTimes(1)
      setTimeout(() => {
        expect(mockSetError).toBeCalledWith('mockError')
        done()
      }, 1000)
    })
  })
})
