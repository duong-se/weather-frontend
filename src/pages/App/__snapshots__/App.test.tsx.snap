// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App should match snapshot 1`] = `
<div
  className="root container-fluid"
>
  <div
    className="card p-4 weatherWidget"
  >
    <div
      className="row no-gutters mb-4 information"
    >
      <LocationSearchForm
        isLoading={false}
        onSubmit={[Function]}
      />
    </div>
    <div
      className="row no-gutters mb-4 information"
    />
    <div
      className="row no-gutters information"
    >
      <ForecastList
        consolidatedWeathers={Array []}
        setCurrentWeather={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`App renderCurrentWeatherInfo should render correctly and have data 1`] = `
<div>
  <div
    className="text-center"
  >
    <h1>
      Heavy Rain
    </h1>
    <div>
      <span
        className="temperatureText"
      >
        32
      </span>
      <span>
        â„ƒ
      </span>
    </div>
    <h4>
      Ho Chi Minh City
    </h4>
    <h6>
      Thursday
    </h6>
  </div>
</div>
`;

exports[`App renderCurrentWeatherInfo should render correctly and no data 1`] = `<div />`;

exports[`App renderError should render null 1`] = `<div />`;

exports[`App renderError should renderError 1`] = `
<div>
  <div
    className="row no-gutters mb-4"
  >
    <Alert
      message="MockError"
    />
  </div>
</div>
`;
